cache_flat_mutation_reader.hh

```diff
@@ -642,10 +657,38 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
     }, _read_context.is_reversed()) == stop_iteration::no) {
         return;
     }
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
+
     if (_next_row_in_range) {
-        add_to_buffer(_next_row);
+        bool remove_row = false;
+
+        if (_state == state::reading_from_cache && !_next_row.dummy()) { // maybe we'd better use at_a_row() instead of dummy() here?
+            deletable_row& row = _next_row.latest_row();
+            tombstone base_tombstone = _snp->version()->partition().range_tombstone_for_row(*_schema, _next_row.key());
+
+            if (!row.is_live(*_schema, column_kind::regular_column, base_tombstone, _read_time) && // should we really call is_live here?
+                row.deleted_at().max_deletion_time() < _gc_before) {
+                remove_row = true;
+            }
+        }
+
+        if (remove_row) {
+            with_allocator(_snp->region().allocator(), [&] {
+                partition_snapshot_row_weakref row_ref(_next_row);
+
+                cache_tracker& tracker = _read_context.cache()._tracker;
+                if (row_ref->is_linked()) {
+                    tracker.get_lru().remove(*row_ref);
+                }
+                row_ref->on_evicted(tracker);
```

Comment:
You cannot evict from latest version if there are older MVCC versions, see https://github.com/scylladb/scylladb/blob/master/docs/dev/row_cache.md:

```
Another rule is that row entries in older versions are evicted first, before any row in the newer version is evicted. This is needed so that we don't appear to loose writes in case we have the same row in more than one version, and the one from the newer version gets evicted first. To achieve this, we only move to the front of the LRU (marking as more recently used, evicted last) row entries which belong to the latest partition_version in a given partition_entry. This implies that detached snapshots never update the LRU.
```

There is already `cache_flat_mutation_reader::maybe_drop_last_entry()`, which checks for sufficient conditions for removal. Can we extend that?

Reply:
Thanks, will check the subject